### PR TITLE
Wire Rust resolver into LSP autocomplete and fix completion bugs

### PR DESCRIPTION
**Pull request by dqnykamp**
## Summary

Wire the Rust WASM resolver (`PublicDoenetMLCore.resolve_path`) into the LSP autocomplete pipeline, replacing the JS-only heuristic for `$name.member` resolution. Then fix a series of autocomplete bugs and add several new features on top of the resolver integration.

### Rust resolver integration (commit 1)
- `RustResolverAdapter` bridges the `DoenetSourceObject` DAST with the Rust WASM core, maintaining synchronized index mappings
- Replaces the JS fallback for `resolveRefMemberContainerAtOffset` and adds `isNameAddressableFromOffset` for scope-aware `$name` filtering
- Lazy WASM initialization in the LSP server (`validate.ts` / `rust-core.ts`)

### Hardening (commit 2)
- Guard `syncSource()` and `buildMappings()` against empty, text-only, and incomplete DAST sources
- Regression tests for edge cases (empty doc, `<a`, text-only, recovery after blank)

### Autocomplete bug fixes and features (commit 3)

**Sugar visibility (Bugs A & B from audit)**
- `isHiddenBySugar()` post-filter: names inside sugared `conditionalContent`/`select` are hidden from outside their scope
- Transparent walking through `case`/`else`/`option` wrappers for `$cc.member` and `$sel.member` completions

**`takesIndex` support**
- Add `static takesIndex = true` to 17 composite component classes (Repeat, Select, Sequence, etc.)
- Wire `takesIndex` through schema generation → `doenet-schema.json` → adapter → completion items
- Suppress dot-access descendants for `takesIndex` composites unless bracket index is present (`$name[1].member`)

**`$name[]` snippet completions**
- Offer `$name[]` snippet items for `takesIndex` elements with cursor positioned between brackets using the CodeMirror `CompletionSnippetCursor` protocol

**Repeat `valueName`/`indexName` injection**
- `getRepeatSyntheticNames()` injects `valueName`/`indexName` as completions inside repeat bodies
- Also available as member completions via `$rep[1].valueName`

**Index bracket parsing**
- Context parser recognizes `$name[n]` syntax, strips indices from path parts, and threads `hasIndex` through the resolver and completion pipeline

**Partial resolution**
- Unresolved intermediate path segments (`$s.nonexistent.`) return `node: null` → no completions (previously leaked parent properties)

**Vite/Cypress fix**
- Custom `raw-lsp-bundle` plugin (base64 encoding) to handle `?raw` import of the ~7 MB LSP IIFE bundle with inlined WASM

## Tests
- **133 vitest** tests pass (31 new integration tests)
- **42 Cypress component** tests pass (1 new `$name[]` snippet cursor test)
- All files formatted with Prettier

### CODE REVIEW

File: .changeset/solid-radios-laugh.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+"@doenet/doenetml": patch
+"@doenet/standalone": patch
+"@doenet/doenetml-iframe": patch
+---
```
Comment (siefkenj) [inline .changeset/solid-radios-laugh.md:5]: What's with the name of this file? Make it appropriate.

File: packages/codemirror/src/extensions/lsp/worker.ts [hunk 2]
```diff
@@ -1,5 +1,12 @@
-// @ts-ignore
-import LSPWorker from "@doenet/lsp/language-server.js?worker";
+// @ts-ignore — ?raw loads the pre-built IIFE as a string so we can create
+// a blob: URL Worker.  This avoids the double-base64 encoding that Vite's
+// ?worker inline produces (data: URL of the entire IIFE), which bloated
+// downstream bundles and broke WASM loading inside the worker.
```
Comment (siefkenj) [inline packages/codemirror/src/extensions/lsp/worker.ts:4]: Don't use derogatory terms like "bloat" in comments.

File: packages/codemirror/cypress.config.ts [hunk 3]
```diff
@@ -1,11 +1,46 @@
...
 
 export default defineConfig({
     component: {
         devServer: {
             framework: "react",
             bundler: "vite",
+            viteConfig: {
+                plugins: [
+                    {
+                        // The LSP IIFE bundle (≈7 MB, with inlined WASM)
```
Comment (siefkenj) [inline packages/codemirror/cypress.config.ts]: Put dates on comments if they refer to numbers that might fail to be true in the future.

File: packages/codemirror/vite.config.ts [hunk 4]
```diff
@@ -1,11 +1,38 @@
...
+
+/**
+ * Vite plugin that ensures `?raw` imports of the LSP IIFE bundle work in the
+ * dev server.  The built LSP bundle (≈7 MB, with inlined WASM data URL) can
+ * trip Vite's `import-analysis` transform when the normal `vite:asset` `load`
+ * handler doesn't claim the file (e.g. inside Cypress component-test dev
+ * server).  This plugin intercepts the load early and returns the file as a
+ * plain string export, matching what `?raw` should produce.
+ */
+function rawLspBundlePlugin(): PluginOption {
```
Comment (siefkenj) [inline packages/codemirror/vite.config.ts]: This shouldn't be needed.

File: packages/doenetml-worker-javascript/src/components/BooleanList.js [hunk 5]
```diff
@@ -6,6 +6,8 @@ import { convertUnresolvedAttributesForComponentType } from "../utils/dast/conve
 export default class BooleanList extends CompositeComponent {
     static componentType = "booleanList";
 
+    static takesIndex = true;
+
```
Comment (siefkenj) [inline packages/doenetml-worker-javascript/src/components/BooleanList.js:10]: These are added because they are needed for introspection? It seems weird to have them in this PR.

File: packages/lsp-tools/src/auto-completer/index.ts [hunk 6]
```diff
@@ -128,6 +150,43 @@ export class AutoCompleter {
         return this;
     }
 
+    /**
+     * Replace the name-addressability checker used during `$name` completion.
+     * Pass `undefined` to restore default (allow all names).
+     */
+    setIsNameAddressable(fn?: (offset: number, name: string) => boolean) {
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/index.ts]: Why should this be replaceable?

File: packages/lsp-tools/src/auto-completer/index.ts [hunk 7]
```diff
@@ -128,6 +150,43 @@ export class AutoCompleter {
...
+        this.isNameAddressableImpl = fn;
+        return this;
+    }
+
+    /**
+     * Set a callback that returns additional ref names (e.g. repeat
+     * `valueName`/`indexName`) that should appear in `$name` completions
+     * at the given offset even though they are not in the raw DAST.
+     */
+    setGetAdditionalRefNames(fn?: (offset: number) => string[]) {
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/index.ts]: I don't like these setters. If data needs to be passed in, do it once during initialization. Otherwise, there needs to be a good reason to update these on the fly.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 8]
```diff
@@ -5,13 +5,136 @@ import type {
...
+
+/**
+ * Walk all descendants of `root`, returning an array of `{ name, element }`
+ * for every element with a `name` attribute.
+ */
+function collectAllNamedDescendants(
+    root: DastElement,
+): Array<{ name: string; element: DastElement }> {
+    const result: Array<{ name: string; element: DastElement }> = [];
+    const walk = (node: DastNodes) => {
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: Define functions over variables, unless the function needs to be changed dynamically.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 9]
```diff
@@ -5,13 +5,136 @@ import type {
...
+    }
+
+    return result;
+}
+
+/**
+ * Extract `valueName`/`indexName` attribute values from a repeat or
+ * repeatForSequence element.  Returns an empty array for other elements.
+ */
+function getSyntheticNamesFromElement(el: DastElement): string[] {
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: Maybe `autoGenerated` instead of `Synthetic`?

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 10]
```diff
@@ -44,6 +167,12 @@ interface FlatPathPartForResolver {
 export interface RustResolverAdapterOptions {
     /** An initialized PublicDoenetMLCore (or compatible) instance. */
     core?: RustResolverCore;
+    /**
+     * Optional callback to check whether a component type takes an index.
+     * When provided, the resolver suppresses descendant names for elements
+     * whose type returns `true`, forcing member access through `$name[n].`.
+     */
+    takesIndex?: (componentType: string) => boolean;
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: This is all known statically, right? Let's just get all of this data one time.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 11]
```diff
@@ -59,6 +188,7 @@ export class RustResolverAdapter {
     private core: RustResolverCore | null = null;
     private sourceObj: DoenetSourceObject;
     private enabled = false;
+    private takesIndexFn: ((componentType: string) => boolean) | null = null;
```
Comment (siefkenj) [inline packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: I don't like `private` fields. Make them regular fields and prefix with an `_` if needed.

File: packages/lsp/src/features/validate.ts [hunk 12]
```diff
@@ -59,15 +64,63 @@ export function addValidationSupport(
...
+        if (info.rustAdapter) {
+            // Adapter already wired — just resync source.
+            info.rustAdapter.updateSource(info.autoCompleter.sourceObj);
+        } else {
+            // Fire-and-forget: first completions use the JS fallback until
+            // the WASM module finishes loading, then the Rust resolver
+            // activates for subsequent requests.
+            const capturedInfo = info;
+            getRustCore()
+                .then((core) => {
```
Comment (siefkenj) [inline packages/lsp/src/features/validate.ts]: Make this callback `async` instead of doing old-style promise chains.

File: packages/lsp/src/rust-core.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+ * The LSP keeps one adapter/core per open document so Rust-side source state
+ * and JS-side index mappings stay aligned in multi-document sessions.
+ */
+export function getRustCore(): Promise<PublicDoenetMLCore> {
+    return (async () => {
+        await ensureRustWasmInitialized();
+        const core = PublicDoenetMLCore.new();
+        // Flags must be set before the core can process source.
+        // An empty flags object is sufficient for path resolution.
+        core.set_flags("{}");
```
Comment (siefkenj) [inline packages/lsp/src/rust-core.ts]: Isn't this `async`?

File: .changeset/solid-radios-laugh.md (hunk 1 above)
Comment (dqnykamp) [reply .changeset/solid-radios-laugh.md:5]: This is automatically generated by the `changeset` algorithm. :smile:  It is then deleted in the publish PR #977. It appears to just choose three random words to avoid collisions.

File: packages/doenetml-worker-javascript/src/components/BooleanList.js (hunk 5 above)
Comment (dqnykamp) [reply packages/doenetml-worker-javascript/src/components/BooleanList.js:10]: They are needed because I need a unique marker to designate which components should include autocompletes with array indices. Given some idiosyncrasies in Doenet's ref resolution, I couldn't infer which components should have indices by analyzing the other variables. I had to manually decide and record the decision somewhere. I think `conditionalContent` ended up being the oddball. It may have been the only composite component that should not take indices. An alternative approach could be to just mark conditionalContent (and others, if they exist) as being the oddball in that it doesn't take indices despite being a composite.

It is currently only used for autocomplete. I could also make the attribute name specific for that.

File: packages/codemirror/vite.config.ts (hunk 4 above)
Comment (dqnykamp) [reply packages/codemirror/vite.config.ts]: Thanks for catching this. Indeed, this appears to be AI's first attempt to fix the component tests, and they didn't revert it after finding the needed fix in cypress.config.ts.

File: packages/lsp-tools/src/auto-completer/index.ts (hunk 6 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/index.ts]: The LSP had both a JS resolver and a Rust resolver, and it was falling back to the JS for auto-complete if the rust was not available. Now, it is refactored to only use the rust resolver so the autocompletion is consistent (and returns no completions if the rust wasm is not initialized).

File: packages/lsp-tools/src/auto-completer/index.ts (hunk 7 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/index.ts]: An AI drafted comment for fun:

Good point—I removed the mutable setter APIs and now initialize AutoCompleter with all resolver dependencies at construction time, replacing the instance when Rust becomes ready instead of mutating it in place.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts (hunk 9 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: These aren't auto-generated names, these are names from when an author types `<repeat valueName="v" itemName="i">` and then can use `$v` and `$i` to reference the value and index of the the repeat.

We'll try `derivedRepeatNames`.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts (hunk 8 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: updated

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts (hunk 10 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: precomputed `takesIndex`

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts (hunk 11 above)
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: done

File: packages/lsp/src/features/validate.ts (hunk 12 above)
Comment (dqnykamp) [reply packages/lsp/src/features/validate.ts]: done

File: packages/lsp/src/rust-core.ts (hunk 13 above)
Comment (dqnykamp) [reply packages/lsp/src/rust-core.ts]: made it async

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 14]
```diff
@@ -44,54 +176,83 @@ interface FlatPathPartForResolver {
...
+    _sourceObj: DoenetSourceObject;
+    _enabled = false;
+    _takesIndexComponentTypes: ReadonlySet<string> | null = null;
 
     /** Rust flat index → JS DAST element (matched by source position). */
-    private rustIndexToDastElement: Map<number, DastElement> = new Map();
+    _rustIndexToDastElement: Map<number, DastElement> = new Map();
     /** JS DAST element → Rust flat index. */
-    private dastElementToRustIndex: Map<DastElement, number> = new Map();
+    _dastElementToRustIndex: Map<DastElement, number> = new Map();
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: Don't want to add private function, as we aren't making a library for consumption, and that would make debugging code in the console challenging. We did add `readonly` to some.

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 15]
```diff
@@ -44,57 +176,103 @@ interface FlatPathPartForResolver {
...
     }
 
+    private componentTakesIndex(componentType: string): boolean {
+        return this._takesIndexComponentTypes?.has(componentType) ?? false;
+    }
+
     /**
      * Sync the DAST/source to the Rust core and rebuild index mappings.
      */
     private syncSource(): void {
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 16]
```diff
@@ -121,102 +299,422 @@ export class RustResolverAdapter {
...
+                        );
+                    });
+                if (blockedPathPartIndex >= 0) {
+                    return {
+                        node: null,
+                        unresolvedPathParts:
+                            lookupParts.slice(blockedPathPartIndex),
+                    };
+                }
+            }
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:507]: fixed

File: packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts [hunk 17]
```diff
@@ -10,7 +10,29 @@ type MacroNode = DastMacro;
...
+    let hasIndex = false;
+    const stripped = parts.map((p) => {
+        if (BRACKET_INDEX_SUFFIX_REGEX.test(p)) {
+            hasIndex = true;
+            return p.replace(BRACKET_INDEX_SUFFIX_REGEX, "");
+        }
+        return p;
+    });
+    return { parts: stripped, hasIndex };
+}
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts:38]: fixed

File: packages/codemirror/src/extensions/lsp/plugin.ts [hunk 18]
```diff
@@ -225,6 +225,9 @@ export class LSPPlugin implements PluginValue {
 
     destroy() {
         this.unsubscribeDiagnostics?.();
+        void uniqueLanguageServerInstance
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/codemirror/src/extensions/lsp/plugin.ts]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 19]
```diff
@@ -121,102 +296,437 @@ export class RustResolverAdapter {
...
+                        this._componentTakesIndex(pathNode.name) &&
+                        !segmentHasIndex
+                    ) {
+                        return {
+                            node: null,
+                            unresolvedPathParts:
+                                lookupParts.slice(pathPartIndex),
+                        };
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:506]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 20]
```diff
@@ -121,102 +296,444 @@ export class RustResolverAdapter {
...
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:506]: fixed

File: packages/lsp/src/rust-core.ts [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+        if (normalized.startsWith("file://")) {
+            const fromFileUrl = decodeURIComponent(
+                new URL(normalized).pathname,
+            );
+            if (fromFileUrl) {
+                candidatePaths.push(fromFileUrl);
+            }
+        } else if (normalized.startsWith("/@fs/")) {
+            candidatePaths.push(normalized.replace(/^\/@fs\//, "/"));
+        } else if (path.isAbsolute(normalized)) {
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp/src/rust-core.ts:100]: fixed

File: packages/lsp/src/features/completions.ts [hunk 22]
```diff
@@ -11,6 +11,20 @@ export function addDocumentCompletionSupport(
...
+        if (
+            isRustDependentRefContext &&
+            (info.rustState !== "ready" || !info.rustAdapter)
+        ) {
+            return [];
+        }
+
         const completions = info.autoCompleter.getCompletionItems(
             params.position,
         );
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp/src/features/completions.ts:31]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 23]
```diff
@@ -1,17 +1,149 @@
...
+                    nameAttr.children?.length === 1 &&
+                    nameAttr.children[0].type === "text"
+                        ? nameAttr.children[0].value
+                        : undefined;
+                if (nameVal) result.add(nameVal);
+            }
+            addUniqueNamesFromDescendants(
+                collectAllNamedDescendants(child),
+                result,
+            );
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:130]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 24]
```diff
@@ -1,17 +1,172 @@
...
     ResolveRefMemberContainerArgs,
 } from "./index";
 
+/**
+ * Wrapper elements whose children should be treated as direct children of
+ * the composite for autocomplete purposes.  At runtime, sugar inserts these
+ * wrappers and then strips them during replacement expansion.
+ */
+const COMPOSITE_WRAPPER_NAMES = new Set(["case", "else", "option"]);
+
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:15]: fixed

[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:15]: Oops, this was an incorrect fix. Component types are case-sensitive, so we should not have a case-insensitive match. Reverted the changes made in reponses to this comment.

File: packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts [hunk 25]
```diff
@@ -320,12 +362,7 @@ export function getCompletionContext(
             const pathStr = source.slice(pathStart, offset).trim();
             const pathParts = pathStr.split(".").filter((p) => p.length > 0);
             if (pathParts.length > 0 && /^[A-Za-z0-9_]/.test(pathStr)) {
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts]: fixed

File: packages/doenetml-worker-rust/lib-doenetml-core/src/dast/ref_resolve/resolve.rs [hunk 26]
```diff
@@ -61,7 +61,8 @@ pub(super) enum Visibility {
     ChildrenInvisibleToTheirGrandparents,
 }
 
-const CHILDREN_INVISIBLE_TO_THEIR_GRANDPARENTS: [&str; 3] = ["repeat", "option", "case"];
+const CHILDREN_INVISIBLE_TO_THEIR_GRANDPARENTS: [&str; 5] =
+    ["repeat", "repeatForSequence", "option", "case", "else"];
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/doenetml-worker-rust/lib-doenetml-core/src/dast/ref_resolve/resolve.rs:65]: This case-sensitivity is by design. Component types are case-sensitive. `<repeatForSequence>` is a valid DoenetML tag. `<repeatforsequence>` is an invalid tag.

File: packages/lsp-tools/src/auto-completer/methods/get-completion-items.ts [hunk 27]
```diff
@@ -402,22 +483,44 @@ export function getCompletionItems(
...
-        const propertyNames =
-            this.schemaElementsByName[componentType]?.properties?.map(
-                (property) => property.name,
-            ) || [];
+        const schema = this.schemaElementsByName[componentType];
+        const isDirectRef = completionContext.pathParts.length <= 2;
+        const takesIndex = schema?.takesIndex ?? false;
+        const directPartHasIndex =
+            (completionContext.pathPartHasIndex?.[0] ?? false) || false;
+
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/methods/get-completion-items.ts]: fixed

File: packages/lsp/src/rust-core.ts [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+        const normalized = stripQueryAndHash(wasmBlobUrl);
+        const candidatePaths: string[] = [];
+
+        if (normalized.startsWith("file://")) {
+            const fromFileUrl = decodeURIComponent(
+                new URL(normalized).pathname,
+            );
+            if (fromFileUrl) {
+                candidatePaths.push(fromFileUrl);
+            }
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp/src/rust-core.ts:94]: fixed

File: packages/lsp-tools/src/auto-completer/methods/get-completion-items.ts [hunk 29]
```diff
@@ -375,18 +415,61 @@ export function getCompletionItems(
...
+        // Inject additional names (e.g. repeat valueName/indexName) that
+        // don't exist in the raw DAST but are valid at runtime.
+        const additionalNames = this.getAdditionalRefNames(offset);
+        const uniqueNames = [
+            ...new Set([...addressableNames, ...additionalNames]),
+        ];
+        const filteredNames = uniqueNames.filter(
+            (name) =>
+                this.isNameAddressable(offset, name) &&
+                (!prefix || name.toLowerCase().startsWith(prefix)),
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/methods/get-completion-items.ts]: fixed

File: packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts [hunk 30]
```diff
@@ -44,69 +214,115 @@ interface FlatPathPartForResolver {
...
+            this._enabled = false;
+            return;
+        }
+        // The Rust core panics on empty source (index-out-of-bounds on a
+        // zero-length collection).  It also panics when the DAST contains
+        // no elements (e.g. source is just "a" — only text nodes, zero
+        // elements).  Skip the sync in both cases.
+        if (!this._sourceObj.source.trim()) {
+            this._enabled = false;
+            return;
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/rust-resolver-adapter.ts:282]: fixed

File: packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts [hunk 31]
```diff
@@ -35,8 +60,31 @@ export type CompletionContext =
...
+    rawPathParts: string[],
+): CompletionContext & { cursorPos: "refMember" } {
+    const { parts, pathPartHasIndex } = stripIndicesFromPathParts(rawPathParts);
+    return {
+        cursorPos: "refMember",
+        typedPrefix,
+        replaceFromOffset,
+        pathParts: parts,
+        pathPartHasIndex,
+    };
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/methods/get-completion-context.ts:85]: Spaces in references are not valid. Modified the autocomplete code to make that clearer.

File: packages/lsp-tools/src/auto-completer/index.ts [hunk 32]
```diff
@@ -30,21 +35,32 @@ export type ResolveRefMemberContainerArgs = {
...
 export type RefMemberContainerResolution = {
     node: DastElement | null;
     unresolvedPathParts: string[];
+    /**
+     * When provided by the resolver, this is the list of descendant names
+     * that are actually visible from the resolved node (respecting
+     * visibility rules like `ChildrenInvisibleToTheirGrandparents`).
+     * The completion pipeline expects this to be provided by the resolver.
+     */
+    visibleDescendantNames?: string[];
```
[the remark this answers is not in the document]
Comment (dqnykamp) [reply packages/lsp-tools/src/auto-completer/index.ts]: fixed
